### ISSUE TITLE
Option to force Poetry to create a virtual environment, even if a virtual env is active

### ISSUE THREAD

**Issue by chernals** [python-poetry/poetry#4055]
I am using Poetry from within a `conda` environment; with Poetry being installed by conda (`poetry` is present in the `environment.yaml` file). This might not be ideal but for a specific setup this seems to work well.

When running `poetry install`, a `venv` is not created because Poetry detect that a virtual environment (the Conda one) is already active.

I believe that we should be able to force Poetry to create a virtual environment anyway: the Conda environment is more "physical" than virtual in this case, as it replaces the system Python.

The same issue occurs when doing `poetry shell` from within the Conda environment, after creating the virtual environment with the following workaround: I deactivate the Conda environment, use the full path to Poetry (in the `bin` directory of the Conda environment) and do `poetry install`.

**Comment by chernals** [comment]
This is related to ISSUE-1724 and to ISSUE-4050, but in the form of an explicit request to add an option.

**Comment by mforbes** [comment]
What would really be nice is a way to have an active underlying conda environment upon which poetry could create a virtual environment for installing dependencies while maintaining access to the underlying conda environment.

A use-case is a group of students working on a shared computer (like an HPC cluster): I would like to create a fairly full featured shared conda environment (called `work` here) which students have read access to, but not write.  Currently poetry will treat this active conda environment as a virtual environment, but fail because it has no write access.

The behaviour I would like (if possible) is for poetry to be able to install or upgrade any missing dependencies as needed, akin to how `pip install --user` would, but to install these into a custom virtual environment rather than the users `~/.local`.

I know that I could do this by manipulating paths `PYTHONUSERBASE` etc. but it would be really nices if this could "just work" in a way that consistent with general poetry usage.

Thus, daily work looks like:

```bash
conda activate work  # Activate read-only conda environment
poetry shell         # Creates a new virtual environment with access to `work`
python ...           # Works for anything using just `work`
poetry add ...       # If something is not in `work` (currently fails because of permissions)
poetry install       # Installed into the virtual environment like pip install --user
```

This allows students to get to work as quickly as possible, allowing us to provide most
of what they need in the `work` environment, but providing them a way to install other
libraries if needed.  Everything new needed is recorded in `pyproject.toml`, so in
combination with the environment file for the `work` environment, this satisfies the
basic requirements for reproducibility.

When preparing for release, one would add the minimal set of missing dependencies
supplied by work into a clean virtual environment:

```bash
poetry env use 3.8
poetry shell         # Clean environment without `work` 
poetry add ...       # Clean up all dependences for release.
...
```

Now students need to work harder to properly define a clean minimal set of dependencies,
but this task does not stand in the way of them getting started.

Any suggestions for alternatives would be appreciated.

**Comment by mforbes** [comment]
One other feature that would be really nice is if `poetry shell` could activate whatever environment is needed, including calling `conda activate`.  I am not sure how internally this works, but the following workflow would be very useful:

```bash
conda activate my_conda_env
poetry env use python        # Tell poetry to use the my_conda_env
poetry shell                 # This works.
```

After doing this, it would be very nice if this or something similar could store information about the choice of environment in `poetry.toml` or something so that running `poetry shell` from within the project first activates the `my_conda_env` (until `poetry env use` or similar is called in the future).

**Comment by reitzig** [comment]
I don't use conda, but I second this feature request anyways.

In trying to debug a failing CI pipeline, it helps a lot if the venv is exclusive to the current build. On Jenkins, that can only be assured by creating a venv in the current workspace, _no matter_ which environments exist.

Currently, I have no way (?) to force poetry to _not_ use an existing environment (which I can't delete), which may or may not be the root of my problem.

**Comment by thomkeh** [comment]
I am also missing this feature, as I am migrating some code from `pipenv` to `poetry`. `pipenv` has `PIPENV_IGNORE_VIRTUALENVS` which has [exactly the effect](https://stackoverflow.com/questions/48680012/force-pipenv-to-create-a-new-virtualenv) that's wished for here.

**Comment by JayZar21** [comment]
I described a workaround in this answer:

https://stackoverflow.com/questions/70739858/how-to-create-a-brand-new-virtual-environment-or-duplicate-an-existing-one-in-po

However, it would be nice to have the possibility to define more pyproject.toml and create other environments with the framework.

**Comment by tony** [comment]
This makes sense. I've had this happen to me again today

`pipenv(1)` does this behavior

IMO, In this case this would still be valid poetry usage, even if poetry wasn't necessarily installing the packages.

**Comment by popadi** [comment]
Any more update/traction on this? 

My usecase is upgrading some projects from py3.7 to py3.10 and it's really annoying having two conda envs for each project. It will greatly help forcing poetry to create a local env every time.

**Comment by Aperyon** [comment]
Deleting the existing virtualenv directory did help me.

```bash
$ poetry run python -m site
sys.path = [
    ...
    '/Users/user-name/Library/Caches/pypoetry/virtualenvs/some-project/lib/python3.10/site-packages',
]
```

After this I just simply ran

```bash
poetry install
```

and a new virtualenv env was automatically created.

I also added these lines to the project's `poetry.toml` (even though these are my global configs as well)

```
[virtualenvs]
create = true
in-project = true
```

**Comment by jaklan** [comment]
+1 for the `conda` case, it's pretty problematic in our remote environment, where `python` interpreter is provided by `conda` and `poetry install` doesn't create project-specific venvs...

**Comment by Green-li** [comment]
> +1 for the `conda` case, it's pretty problematic in our remote environment, where `python` interpreter is provided by `conda` and `poetry install` doesn't create project-specific venvs...

I hava found [PDM](https://github.com/pdm-project/pdm), which meets my requirements.

**Comment by coccoinomane** [comment]
>I have found [PDM](https://github.com/pdm-project/pdm), which meets my requirements.

Wow, PDM is like `npm` for Python... thanks for sharing!

**Comment by Dave-Vallance** [comment]
Vote for this issue please. 

My use case is that I have a program which will checkout projects and install them in their own virtual environment. Some of the projects we are checking out use poetry. 

The program which is checking out the repos and doing the installs also runs from its own virtual environment. The result appears to be that all the project dependencies end up in the virtual environment of the running program if that makes sense? It would be great to allow poetry to still create virtual environment and install to them.

**Comment by renan-r-santos** [comment]
I faced this problem as well and created a [Poetry plugin](https://github.com/renan-r-santos/poetry-conda) that fixes this issue for the Conda use case. Although not ideal, this solution seems to work well when I tested it.
Would Poetry maintainers be open to adding a similar config to Poetry? I'd be happy to contribute with a PR.

**Comment by TylerGrantSmith** [comment]
Referencing an earlier comment from @Secrus: https://github.com/python-poetry/poetry/issues/7878#issuecomment-1535496388.  

This current behavior seems very much in contradiction to the linked guidance.  If installing poetry in the same environment is to be discouraged, then attempts to initialize/install a project co-located with poetry should at the very least raise a warning that this is problematic.

**Comment by ndevenish** [comment]
> I faced this problem as well and created a [Poetry plugin](https://github.com/renan-r-santos/poetry-conda) that fixes this issue for the Conda use case

`poetry-conda` has fixed the issue I had with this. Poetry was installed on a (non-virtual) python, but I had a conda environment enabled **that didn't contain python** for some binary tooling (specifically `h5ls` and othe inspection tools). Every time trying to e.g. `poetry run morgul-prepare` it would say
```
The virtual environment found in /path/to/activated/conda/ENV seems to be broken.
Recreating virtualenv morgul-zttW3Bd7-py3.11 in /poetry_cache_path/virtualenvs/morgul-zttW3Bd7-py3.11
Warning: 'morgul-prepare' is an entry point defined in pyproject.toml, but it's not installed as a script. You may get improper `sys.argv[0]`.

The support to run uninstalled scripts will be removed in a future release.

Run `poetry install` to resolve and get rid of this message.

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/dls/science/users/mep23677/modules/python/3.11.3/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1206, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1178, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1149, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/dls/science/users/mep23677/jungfrau/jungfrau-commissioning/morgul/morgul_prepare.py", line 7, in <module>
    import h5py
ModuleNotFoundError: No module named 'h5py'
```
Then run `poetry install`, which appears to succeed, then the `poetry run ....` goes back to the above error message.

Poetry definitely shouldn't be trying to use an existing conda environment that doesn't contain python, and I **never** want it to, so a systemwide setting that lets's me turn off poetry trying to do this would be perfect.

**Comment by gabyx** [comment]
The behavior that poetry always prefers an **acitvated venv is harmful** and leads to stupid workarounds and nonsensical behavior. Normally also normal users have virtual envs on their machine which act as a basis for lets say `python 3.10` which is active in the shell. 

We can only force creating a venv by doing:

```
(unset VIRTUAL_ENV &&  poetry -v install)
```
which will then read the `poetry.toml` which has `virtualenvs.create=true` which then installs into `./.venv` and **not** use the already activated one.

The above workaround however leads to issues with other commands because its not synchronized:
```shell
poetry env info --path
```
will in the same shell not give you the just created , nameley `./.venv` but the system venv.

I a not even get the option `POETRY_VIRTUALENVS_CREATE` in that sense. If you specify `true` it should always create one, full stop. Because the other behavior "it always prefers already activated venvs" leads to really peculiar side effects in CI and on your local machine.

Please fix this in making this easy and comprehensible.

**Comment by maflAT** [comment]
I just tried using poetry for the first time following the introduction guide in the documentation.
The official documentation is very misleading and contradicting itself, which has brought me here!

In the very [first chapter](https://python-poetry.org/docs/#installation) you see this big, red warning:
![grafik](https://github.com/python-poetry/poetry/assets/13979550/a6ade27d-d23b-44d5-aaa5-5cae9ecd70ea)

If you heed this warning, installing poetry in its own environment, and continue with the example you will run into the issue described here in the very next chapter.
Poetry will happily use its own environment for the example project without any question or warning.

[Using your virtual environment](https://python-poetry.org/docs/basic-usage/#using-your-virtual-environment)

> By default, Poetry creates a virtual environment in {cache-dir}/virtualenvs. 

This is just plain wrong at this point. Poetry will not do this if you follow the guide. It will just silently use its own environment.
>You can change the [cache-dir](https://python-poetry.org/docs/configuration/#cache-dir) value by editing the Poetry configuration. 

I have not tried this.
>Additionally, you can use the [virtualenvs.in-project](https://python-poetry.org/docs/configuration/#virtualenvsin-project) configuration variable to create virtual environments within your project directory.

I have not tried this either, but from what I can see in the related issues, this is also wrong.

At this point I started googling why I couldn't find any virtualenvs created by poetry.

Only after that, does the documentation mention that poetry will prefer an already active virtualenv. This might clear things up if you already have some experience or you have googled the issue. But for a first time user this guide is misleading and unusable.

**Comment by radoering** [comment]
@maflAT I understand the confusion if you only read the red box. However, if you read further it does not matter which method you choose (pipx, official installer, manual): all methods make sure that poetry's own virtual environment is not activated. Nevertheless, a PR improving the docs (maybe an additional warning that poetry's own environment should not be activated) is always welcome.

**Comment by aesteve** [comment]
I think the problem does not surface when `poetry` is installed but at a later point in time when used. So I agree, all the methods mentioned here do not activate the environment. But when you actually make use of poetry, say, to run `poetry install` then you shouldn't do that from poetry's virtual environment although it's really tempting to. And actually, the fact the commands mention `poetry install` only could let you think you have activated the virtual environment in which you installed poetry (I actually thought so and was really puzzled by `poetry install` not creating the `.venv` in my sub-directory although `virtualenvs.in-project` was set to true.

I tried to update the warning sign in #8833

**Comment by drhagen** [comment]
With the arrival of [Pixi](https://pixi.sh/latest/), this has become more of a problem. Pixi does not have a base environment by default. If you create a base environment in which to install Python and Poetry, you need to know to call it "base" or else Poetry will not work.

I would recommend this as the solution: Option called `virtualenvs.ignore-conda-envs`. If `true`, Poetry behaves as if it does not see the conda environment at all. If `false`, Poetry always installs into a conda environment when inside one. If a list of strings, Poetry ignores only those environments. The default is `["base"]` keeping the current behavior unchanged.

**Comment by austinmw** [comment]
over 3 yeas later and still no updates??

**Comment by radoering** [comment]
Did you pay attention to the last sentence of the recommendation?

> In addition, the isolated virtual environment in which poetry is installed should not be activated for running poetry commands.

**Comment by radoering** [comment]
> Who and what comment are you responding to?

I responded to the comment directly above my comment but it might also apply to your comment. Respecting/using activated virtual environments is a feature even though it might look like a bug for some users. A perfectly valid workflow is to create a virtual environment without the help of Poetry, activate it and thereby let Poetry use it.

> Also, as a general comment on this issue, I see that it's still labeled as a feature request [...] but as I described above, this is a bug.

Considering the title and the description of this issue I think this issue is a feature request.

**Comment by mforbes** [comment]
@radoering I think that part of the issue some users have is that the instructions say - install poetry in a dedicated virtual environment, but don't activate that environment.  That is not how many people expect virtual environments to work (at least with conda – Pipx makes things better, but condax still has some issues).  While one can manipulate paths to use the tools installed in an isolated environment, this can cause issues if not done carefully (which is the point of activating functional conda environments).

While using an active virtual environment is indeed a feature, it is strange and confusing that there is no robust way to turn off this feature when it causes issues.

Perhaps poetry could refuse to update the environment in which it is installed, and emit a helpful warning at that point of use with appropriate instructions?  Maybe something like:

> It appears you running `poetry` from an activated virtual environment. Any updates will be applied to this active environment, which is probably not what you want. As discussed in the documentation, `poetry` should not be run from an activated environment, but should instead be installed in a a dedicated virtual environment that is not active (but whose `bin/` directory is on your `PATH`).
>
> <instructions on how to force poetry to create a new virtual environment if this feature request is fulfilled:->
>
> Are you sure you want to proceed? (y/N)

**Comment by renan-r-santos** [comment]
The current approach of Poetry treating only the Conda base environment differently
https://github.com/python-poetry/poetry/blob/5bab98c9500f1050c6bb6adfb55580a23173f18d/src/poetry/utils/env/env_manager.py#L290-L293
likely contributes to user confusion, even though there may be valid reasons for this behavior.
To enhance clarity and user control, I think it would be preferable to allow users to explicitly choose whether they want to reuse Conda environments or not. This opt-in/out approach would eliminate the need for Poetry to make assumptions about what users want, thereby providing a more transparent and predictable experience for users.

**Comment by radoering** [comment]
> Perhaps poetry could refuse to update the environment in which it is installed, and emit a helpful warning at that point of use with appropriate instructions? Maybe something like: [...]

Might make sense but you have to consider that asking for user interaction is undesirable in CI.

>It is very difficult, if not impossible, for Poetry users to be able to read the official documentation, configure settings, and then be aware of what Poetry will do.

I understand that the interaction/meaning of the settings can be confusing. Currently, I do not know how to improve the situation. I can only try to explain the intention although I do not know if it will help or add any new information compared to the documentation:

- desired default behavior: always use an activated virtual environment
  This allows you to create a virtual environment to your likings without Poetry and just use it by activating it. If you do not create a virtual environment by yourself and activate it, Poetry does it for you.
- this default behavior cannot be changed at the moment (this feature request is about an option to change this default behavior)
- `virtualenvs.create` defines what to do if no virtual environment has been activated and no environment has been created yet: Create a new virtual environment or use the system environement. The latter is strongly discouraged in most cases.
- `virtualenvs.in-project` defines where to create a virtual environment if no virtual environment has been activated and no environment has been created yet: Create the environment in a `.venv` folder in the project or in the cache directory.
- `virtualenvs.prefer-active-python` defines which Python to use if a virtual environment has to be created.
- `virtualenvs.in-project` and `virtualenvs.prefer-active-python` have no influence on whether a virtual environment is created or not. They only change where/how it is created.

**Comment by austinmw** [comment]
> Did you pay attention to the last sentence of the recommendation?
> 
> > In addition, the isolated virtual environment in which poetry is installed should not be activated for running poetry commands.

pardon my ignorance, but how do you run a poetry command installed in a virtual environment without activating it?

**Comment by ndevenish** [comment]
> > Did you pay attention to the last sentence of the recommendation?
> > > In addition, the isolated virtual environment in which poetry is installed should not be activated for running poetry commands.
> 
> pardon my ignorance, but how do you run poetry installed in a virtual environment without activating it?

You can call the console script created in the `bin/` folder directly without activating. It hardcodes the correct interpreter executable directly.

Tools like pipx take advantage of this and only expose the packages console_scripts (so you don’t get the python of the last environment on PATH. I think there are several other tools that work similarly.

I don’t think this is the primary problem being raised in this issue, however.

**Comment by Turiok** [comment]
Using the same env for conda and poetry destroy my poetry installation in the conda+poetry env.

I have a conda env for COTS like gitleaks, trufflehog or openldap client and poetry.
Then I would like a poetry env for my python dependencies of my code.

I installed my conda env  then poetry install the dependencies in the conda env too.
It bugs me but worked fine until now.

I wanted add a new poetry dependencies with `poetry add <package>`
And poetry uninstall the virtualenv dependency in conda+poetry env.
Then any poetry command failed because virtualenv module was not found.


I'm using poetry-conda plugin now to make it works. But I thing this feature should be in default installation.

**Comment by avaz** [comment]
For my use case would also be very helpful in having the ability to make poetry disregard Conda or any other kind of virtual environment. 
I have a somewhat complex project made of many modules in a mono repo fashion. Each module has its own `pyproject.toml` and it's managed by poetry which means that all of them have their own environment and their libraries don't conflict with other modules libraries. 

This works well but I use something like Conda to install other project level dependencies that aren't necessary python ones, like for instance some dynamic libraries, ODBC drivers, and so on. Conda fits well for this as it allows to have all project dependencies in the same 'context' while maintaining each modules dependencies private to itself. 

On my `environment.yml` I have the python version required by the project as well poetry which helps to keep everything in one place and reproducible. 

In summary, considering the diverse use cases, it is advisable to allow users to determine when to create a new environment or reuse an existing one. For the time being, I am utilizing the [poetry-conda Poetry plugin](https://github.com/renan-r-santos/poetry-conda) to address my requirements. However, it would be beneficial to have this plugin natively supported.

**Comment by asaril** [comment]
I would like to add another use case for a way to always force a poetry venv creation:

Recently, at least on Ubuntu 24.04 using Python 3.12, pip 24.0, pip starts to complain about system-wide installations of packages (which potentially can break the system with conflicts to packages installed using the system package manager).

Their recommended solution is to always use a venv if packages from pip are needed.

Since not all packages I want in my "base" Python are available in apt (conan for example), I created a default venv with the python-venv package (not conda, so I can't use the poetry-conda plugin).
I tried naming the env "base", but it seems this exception is only done for a conda env called base, not one created by venv.

--> There can be installations that have an always-active "base" venv, that is not from conda.
It would be good to either have a configurable variable to set accepted "base venv" names (and not determine it by being from conda), or some flag to force poetry venv creation explicitly (without putting it into the project directory).

**Comment by jasondamour** [comment]
Is there any workaround for virtualenv yet, akin to the conda plugin? All I can think of so far is enforce devs at our company do `alias poetry='env -u VIRTUAL_ENV poetry'`

**Comment by jasondamour** [comment]
The above PR adds the option `virtualenv.force-create` (defaults to `False`), which I think meets all the requirements of this thread. @radoering, your summary comment helped me understand the requirements, would you be willing to review?
